### PR TITLE
region: fix the potential panic . 

### PR DESCRIPTION
**Pull request by bufferflies**
### What problem does this PR solve?

Issue Number: Close #4399 

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Code changes

- Has the configuration change
- Has HTTP API interfaces changed (Don't forget to [add the declarative for the new API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
- Has persistent data change

Side effects

- Possible performance regression
- Increased code complexity
- Breaking backward compatibility

Related changes

- PR to update [`pingcap/docs`](https://github.com/pingcap/docs)/[`pingcap/docs-cn`](https://github.com/pingcap/docs-cn):
- PR to update [`pingcap/tiup`](https://github.com/pingcap/tiup):
- Need to cherry-pick to the release branch

### Release note

```release-note
None.
```

### CODE REVIEW

File: pkg/core/region.go [hunk 1]
```diff
@@ -1004,8 +1011,17 @@ func (r *RegionsInfo) UpdateSubTree(region, origin *RegionInfo, overlaps []*Regi
...
-			r.updateSubTreeStat(origin, region)
-			r.subRegions[region.GetID()].RegionInfo = region
+			// The region tree and the subtree update is not atomic and the region tree is updated first.
+			// If there are two thread needs to update region tree,
+			// t1: thread-A  update region tree
+			// 										t2: thread-B: update region three again
+			//										t3: thread-B: update subtree
+			// t4: thread-A: update region subtree
+			// to keep region tree consistent with subtree, we need to drop this update.
+			if tree, ok := r.subRegions[region.GetID()]; ok {
```
Comment (rleungx) [inline pkg/core/region.go:1021]: Will it lose the update?

Comment (bufferflies) [reply pkg/core/region.go:1021]: no, it keeps the region tree consistency with the sub tree.

File: pkg/core/region_test.go [hunk 2]
```diff
@@ -450,6 +451,17 @@ func TestRegionKey(t *testing.T) {
...
 
+func TestSetRegionConcurrence(t *testing.T) {
+	re := require.New(t)
+	re.NoError(failpoint.Enable("github.com/tikv/pd/pkg/core/UpdateSubTree", `return()`))
+	regions := NewRegionsInfo()
+	region := NewTestRegionInfo(1, 1, []byte("a"), []byte("b"))
+	go func() {
+		regions.AtomicCheckAndPutRegion(region)
+	}()
+	regions.AtomicCheckAndPutRegion(region)
```
Comment (lhy1024) [inline pkg/core/region_test.go:462]: failpoint.Disable

File: pkg/core/region.go [hunk 3]
```diff
@@ -1004,8 +1011,17 @@ func (r *RegionsInfo) UpdateSubTree(region, origin *RegionInfo, overlaps []*Regi
 			// TODO: Improve performance by deleting only the different peers.
 			r.removeRegionFromSubTreeLocked(origin)
 		} else {
-			r.updateSubTreeStat(origin, region)
-			r.subRegions[region.GetID()].RegionInfo = region
+			// The region tree and the subtree update is not atomic and the region tree is updated first.
+			// If there are two thread needs to update region tree,
+			// t1: thread-A  update region tree
+			// 										t2: thread-B: update region three again
```
Comment (HuSharp) [inline pkg/core/region.go:1017]:
```suggestion
			// 										t2: thread-B: update region tree again
```

File: pkg/core/region_test.go (hunk 2 above)
Comment (bufferflies) [reply pkg/core/region_test.go:462]: good catch ,fixed it

File: pkg/core/region.go (hunk 3 above)
Comment (bufferflies) [reply pkg/core/region.go]: fixed

**Review by rleungx — APPROVED** [review]

**Review by lhy1024 — APPROVED** [review]

**Comment by HuSharp** [comment]
please link an issue :)

**Pull request merged**